File: team1-frontend/src/utils/permissions.ts
// src/utils/permissions.ts — Bit Flag RBAC helpers (CCR-017).
// Permission is a number where each bit represents an allowed action:
//   bit 0 = Read  (1)
//   bit 1 = Write (2)
//   bit 2 = Delete(4)
// A user's permission map is `Record<resource, number>` — checked at router
// guard + template level via the authStore.hasPermission() wrapper.

import { Permission, type PermissionAction } from 'src/types/entities';

export { Permission } from 'src/types/entities';
export type { PermissionAction } from 'src/types/entities';

const ACTION_MASK: Record<PermissionAction, Permission> = {
  Read: Permission.Read,
  Write: Permission.Write,
  Delete: Permission.Delete,
};

/**
 * Return true when `perm` includes the bit for `action`.
 *
 * @example
 *   hasPermission(Permission.Read | Permission.Write, 'Read') // true
 *   hasPermission(Permission.Read, 'Delete')                  // false
 */
export function hasPermission(perm: number | undefined, action: PermissionAction): boolean {
  if (perm === undefined || perm === null) return false;
  const mask = ACTION_MASK[action];
  return (perm & mask) === mask;
}

/**
 * Higher-level check against a full permission map keyed by resource.
 *
 * @example
 *   checkResource({ Lobby: 3 }, 'Lobby', 'Write') // true  (3 = Read|Write)
 */
export function checkResource(
  perms: Record<string, number> | undefined,
  resource: string,
  action: PermissionAction,
): boolean {
  if (!perms) return false;
  return hasPermission(perms[resource], action);
}

File: shared/ebs_common/lib/src/auth/permissions.dart
enum PermissionAction { read, write, delete_ }

class Permission {
  static const int read = 1;
  static const int write = 2;
  static const int delete_ = 4;

  Permission._();

  static bool hasPermission(int? perm, PermissionAction action) {
    if (perm == null) return false;
    final mask = switch (action) {
      PermissionAction.read => read,
      PermissionAction.write => write,
      PermissionAction.delete_ => delete_,
    };
    return (perm & mask) == mask;
  }

  static bool checkResource(
    Map<String, int>? perms,
    String resource,
    PermissionAction action,
  ) {
    if (perms == null) return false;
    return hasPermission(perms[resource], action);
  }
}
